File: Python/Scrapping_Scripts/Recherche/Coliving_Recherche.py
import requests
import json
from datetime import datetime, timedelta

def get_old_data(filename):
    try:
        with open(filename, 'r', encoding='utf-8') as file:
            return json.load(file)
    except FileNotFoundError:
        return []

base_url = "https://coliving.com/locations/search-listings"
params = {
    "price_min": "",
    "price_max": "",
    "room_type": "",
    "bathroom": "",
    "location_id": "77",
    "zoom": "4",
    "center[lat]": "46.227638",
    "center[lng]": "2.213749"
}
all_data = []
page = 1
total_scraped = 0

while True:
    params["page"] = page
    response = requests.get(base_url, params=params)

    if response.status_code != 200:
        print(f"Erreur {response.status_code}: {response.text}")
        break

    # Ici, nous utilisons get pour extraire les annonces de la clé "listings"
    data = response.json().get("listings", [])
    
    if not data:  # Si la réponse est une liste vide, nous avons atteint la dernière page
        break

    # Imprimer le nombre d'annonces pour cette page
    print(f"Page {page} : {len(data)} annonces scrapées")

    all_data.extend(data)  # Ajouter les données de cette page à notre liste globale
    total_scraped += len(data)

    if len(data) < 25:  # Si moins de 25 annonces sont renvoyées, nous avons atteint la dernière page
        break

    page += 1

current_date_string = datetime.now().strftime('%Y-%m-%d')
output_file_name = f'../../Resultat_Recherche/Coliving_Recherche/Data_Coliving_Recherche_{current_date_string}.json'

# Lisez les anciennes données
previous_day_string = (datetime.now() - timedelta(days=1)).strftime('%Y-%m-%d')
old_file_name = f'../../Resultat_Recherche/Coliving_Recherche/Data_Coliving_Recherche_{previous_day_string}.json'
old_data = get_old_data(old_file_name)

# Identifiez les nouvelles annonces et les annonces supprimées
new_announcements = [item for item in all_data if item['id'] not in [old_item['id'] for old_item in old_data]]
removed_announcements = [item for item in old_data if item['id'] not in [new_item['id'] for new_item in all_data]]

# Créez la liste des annonces à jour en excluant les annonces supprimées
updated_data = [item for item in all_data if item['id'] not in [removed_item['id'] for removed_item in removed_announcements]]

# Sauvegardez toutes les annonces scrappées du jour
with open(output_file_name, 'w', encoding='utf-8') as file:
    json.dump(all_data, file, ensure_ascii=False, indent=4)

# Sauvegardez les annonces mises à jour
updated_file_name = f'../../Resultat_Recherche/Up_To_Date_Recherche/Updated_Data_Coliving_Recherche_{current_date_string}.json'
with open(updated_file_name, 'w', encoding='utf-8') as file:
    json.dump(updated_data, file, ensure_ascii=False, indent=4)

print(f"Total d'annonces scrapées : {total_scraped}")
print(f"Données du jour sauvegardées dans {output_file_name}!")
print(f"{len(new_announcements)} nouvelle(s) annonce(s).")
print(f"{len(removed_announcements)} annonce(s) supprimée(s).")
print(f"{len(updated_data) - len(new_announcements)} annonce(s) conservée(s).")
print(f"Données mises à jour sauvegardées dans {updated_file_name}!")

File: Javascript/Scrapping_Scripts/Recherche/Coliving_Recherche.js
const axios = require('axios');
const fs = require('fs').promises;
const path = require('path');
const { getCurrentDateString, getPreviousDateString } = require('../dateUtils');
const { getOldData } = require('../dataUtils');

const delay = (duration) => new Promise(resolve => setTimeout(resolve, duration));

const cities = [
    { name: "Paris", id: 339 },
    { name: "Lyon", id: 638 },
    { name: "Marseille", id: 492 },
    { name: "Toulouse", id: 939 },
    { name: "Bordeaux", id: 698 },
    { name: "Nantes", id: 1221 },
    { name: "Rennes", id: 3449 },
    { name: "Lille", id: 659 },
    { name: "Angers", id: 3003 },
    { name: "Grenoble", id: 1050 }
];

const baseURL = "https://coliving.com/locations/search-listings";
let params = {
    price_min: "",
    price_max: "",
    room_type: "",
    bathroom: "",
    initialize: "0",
    zoom: "11",
    "center[lat]": "48.856614",
    "center[lng]": "2.3522219"
};

const scrapeData = async (city) => {
    let allData = [];
    let page = 1;
    let totalScraped = 0;
    params.location_id = city.id;

    while (true) {
        params.page = page;
        const response = await axios.get(baseURL, { params: params });

        if (response.status !== 200) {
            console.log(`Erreur ${response.status}: ${response.statusText}`);
            break;
        }

        const data = response.data.listings || [];
        if (!data.length) {
            break;
        }

        console.log(`Page ${page} : ${data.length} annonces scrapées pour ${city.name}`);
        allData = allData.concat(data);
        totalScraped += data.length;

        if (data.length < 25) {
            break;
        }

        page++;
    }

    const currentDateString = getCurrentDateString();
    const outputFileName = path.join(__dirname, `../../Resultat_Recherche/Coliving_Recherche/Data_Coliving_${city.name}_${currentDateString}.json`);
    const updatedFileName = path.join(__dirname, `../../Resultat_Recherche/Up_To_Date_Recherche/Coliving_Recherche_Up_To_Date/Updated_Data_Coliving_${city.name}_${currentDateString}.json`);

    const previousDateString = getPreviousDateString();
    const oldFileName = path.join(__dirname, `../../Resultat_Recherche/Coliving_Recherche/Data_Coliving_${city.name}_${previousDateString}.json`);

    const oldData = getOldData(oldFileName);
    const newAnnouncements = allData.filter(item => !oldData.some(oldItem => oldItem.id === item.id));
    const removedAnnouncements = oldData.filter(oldItem => !allData.some(item => item.id === oldItem.id));
    const updatedData = allData.filter(item => !removedAnnouncements.some(removedItem => removedItem.id === item.id));

    await fs.writeFile(outputFileName, JSON.stringify(allData, null, 4), 'utf-8');
    await fs.writeFile(updatedFileName, JSON.stringify(updatedData, null, 4), 'utf-8');

    console.log(`Total d'annonces scrapées pour ${city.name} : ${totalScraped}`);
    console.log(`Données du jour pour ${city.name} sauvegardées dans ${outputFileName}!`);
    console.log(`TOTAL_NOUVELLES_ANNONCES pour ${city.name}: ${newAnnouncements.length} nouvelles annonces.`);
    console.log(`${removedAnnouncements.length} annonce(s) supprimée(s) pour ${city.name}.`);
    console.log(`${updatedData.length - newAnnouncements.length} annonce(s) conservée(s) pour ${city.name}.`);
    console.log(`Données mises à jour pour ${city.name} sauvegardées dans ${updatedFileName}!`);
};

const scrapeCitiesWithDelay = async () => {
    for (const city of cities) {
        await scrapeData(city).catch(error => console.error(`Erreur lors du scraping de ${city.name}: `, error));
        await delay(60000); // Délai de 1 minute
    }
};

scrapeCitiesWithDelay();
